Appendix
============

Make mappability and gap files
------------------------------------------

DROMPA accepts the mappability files and gap files generated by scripts provided by MOSAiCS
[4], which is based on the code from Peakseq [3].
Using the MOSAiCS scripts with the appropriate fragment length and binsize (here 150 bp
and 100 bp, respectively), you can obtain the following files:
- binary mappability files “chr* map binary.txt”;
- bin-level mappability files “chr* map fragL150 bin100.txt”;
- gap files “chr* N binary.txt”.

See the MOSAiCS website 9 for more details.
The bin-level mappability files are used for the “-mp” option of parse2wig, drompa peakcall and drompa draw while the binary files are used for the “-mpbin” option in GC normalization (see section 5.6).

After generateing those files, change their names for DROMPA::

  $ for i in $(seq 1 22) X Y M; do
  $ mv chr${i}_map_binary.txt map_chr${i}_binary.txt
  $ mv chr${i}_map_fragL150 bin100.txt map_fragL150_chr${i}_bin100.txt
  $ mv chr${i}_N_fragL150 bin100.txt N_fragL150_chr${i}_bin100.txt
  $ done

After this, make the “mappability table”, a tab-delimited file describing the number of mappable bases for each chromosome, using makemappabilitytable.pl in the scripts directory.::
  
  # specify the prefix of binary mappability files for the second arguments
  $ scripts/makemappabilitytable.pl genome_table.txt map > map fragL150 genome.txt

The prefix of the “mappability table” should be identical to the bin-level mappability files.

Finally, DROMPA can adopt these files as follows::
  $ -mp map_fragL150 # means map_fragL150_chr*_bin100.txt
  $ -mpbin map # means map_chr*_binary.txt
  $ -gap N_fragL150 # means N_fragL150_chr*_bin100.txt

Gene-density files
----------------------

Gene density files can be generated through the makegenedensity.pl in the scripts directory. To
use a 500 kbp window, type::

  $ scripts/makegenedensity.pl genometable.txt refFlat.txt 500000

and the gene-density files “chr*-bs<binsize>” in the current directory. These data for several
species can also be downloaded from the DROMPA website. Next make a new directory for
gene-density files::
  
  $ mkdir gene_density_hg19
  $ mv chr*-bs* gene_density_hg19

and the gene-density files can be specified as follows::
  
  $ drompa_draw GV $s1 $s2 $s3 $s4 -p ChIPseq-wholegenome -gt genometable.txt \
  $ -GD gene_density_hg19/ -gdsize 500000
